lib/native/source/libboolector/interface_wrap.c

```diff
@@ -3654,137 +3652,157 @@ SWIGEXPORT jobjectArray JNICALL Java_org_sosy_1lab_java_1smt_solvers_boolector_B
-  
+
+  (*jenv)->SetObjectArrayElement(jenv, jniArray, 1, (*jenv)->NewStringUTF(jenv, (const char *)fileOutString));
+
   (*jenv)->SetObjectArrayElement(jenv, jniArray, 2, (*jenv)->NewStringUTF(jenv, errormsg));
-  
+
   (*jenv)->SetObjectArrayElement(jenv, jniArray, 3, (*jenv)->NewStringUTF(jenv, statusString));
-  
+
   (*jenv)->SetObjectArrayElement(jenv, jniArray, 4, (*jenv)->NewStringUTF(jenv, flagString));
 
   (*jenv)->DeleteLocalRef(jenv, classString);
+  unlink(tempfileNameOut);
+  fclose(fileOut);
   free(buffer);
-  unlink(tfnwdIn);
-  free(tfnwdIn);
-  fclose(fparse);
-  unlink(tfnwdOut);  
-  free(tfnwdOut);
-  fclose(fout);
   free(statusString);
   free(resultString);
   return jniArray;
 }
-    
+
 //dumps NODE into new file and reads it to give it back
 SWIGEXPORT jstring JNICALL Java_org_sosy_1lab_java_1smt_solvers_boolector_BtorJNI_boolector_1help_1dump_1node_1smt2(JNIEnv *jenv, jclass jcls, jlong jarg1, jlong jarg2) {
   jstring jresult = 0;
   Btor *arg1 = (Btor *) 0 ;
-  char *result = 0 ;
-  char filename[] = "boolector_help_dump_node_smt2_tempinfile-XXXXXX";
-  FILE *f = 0;
-  char * buffer = 0;
-  int fd = -1;
-  long length = 0;
+  char filenameTemplate[] = "boolector_help_dump_node_smt2_tempinfile-XXXXXX";
+  FILE *file = 0;
+  char *buffer = NULL;
+  int fileDesrc = -1;
+  long fileLength = 0;
   BoolectorNode *arg2 = (BoolectorNode *) 0 ;
-  char *tfnwd = addTemppathToFilename(filename);
+  char *tempfileName = addTemppathToFilename(filenameTemplate);
 
-  fd = mkstemp(tfnwd);
-  if(fd == -1) {
-    free(tfnwd);
-    perror("ERROR CREATING TEMPORARY FILE FOR");
+  if (tempfileName == NULL) {
+    perror("ERROR CREATING TEMPORARY FILE FOR BOOLECTOR_HELP_DUMP_NODE_SMT2");
+    SWIG_JavaThrowException(jenv, SWIG_JavaNullPointerException, "FileName for boolector_help_dump_node_smt2 may not be NULL");
+    return 0;
+  }
+
+  fileDesrc = mkstemp(tempfileName);
+  if (fileDesrc == -1) {
+    free(tempfileName);
+    perror("ERROR CREATING TEMPORARY FILE FOR BOOLECTOR_HELP_DUMP_NODE_SMT2");
     SWIG_JavaThrowException(jenv, SWIG_JavaNullPointerException, "FileDescriptor for boolector_help_dump_node_smt2 may not be NULL");
     return 0;
   }
-    
+
   (void)jenv;
   (void)jcls;
   arg1 = *(Btor **)&jarg1;
-  arg2 = *(BoolectorNode **)&jarg2; 
+  arg2 = *(BoolectorNode **)&jarg2;
 
-  f = fdopen(fd, "w+");
-  if(f==NULL) {
-    free(tfnwd);
-    unlink(tfnwd);
-    perror("ERROR: COULDNT DUMP NODE BECAUSE IT COULDNT CREATE A DUMP FILE"); 
+  file = fdopen(fileDesrc, "w+");
+  unlink(tempfileName);
+  if (file == NULL) {
+    perror("ERROR: COULDNT DUMP NODE BECAUSE IT COULDNT CREATE A DUMP FILE");
     SWIG_JavaThrowException(jenv, SWIG_JavaNullPointerException, "File for boolector_help_dump_node_smt2 may not be NULL");
     return 0;
   }
-    
+
   //write
-  boolector_dump_smt2_node(arg1, f, arg2);
-    
+  boolector_dump_smt2_node(arg1, file, arg2);
+  rewind(file);  //Just to be sure
```

Comment:
Is rewind really needed here?
Do we need to flush the file after boolector has written to it?

Reply:
I am honestly not sure if `rewind()` is needed here. It worked fine without but i wanted to be sure that the stream is at the beginning of the file.
I can take another look at the Boolector code to determin if its really needed though.

Reply:
No we don't need to flush the file as it is deleted after we close the file (tempfile).
Every call to this method should give us a completely new and empty file to write to.

Reply:
But we `read` from it. Is content, which might not yet be flushed, already available for reading?

Reply:
Shouldn't that be the job of the system (giving us an empty file)?
But thats easy to test^^

Reply:
Boolector dumps the String into a file, then we flush (or do currently not flush), and then we read the file into a Java-String. The question is whether we need to flush in the middle or not.

Reply:
Yes but a temporary file that is newly created every time with a unique filedescriptor. If mkstemp can't create a unique firedescr. it gives back -1 and we check that. So it should not be the cause.

Reply:
There might be a misunderstanding: I am not talking about the tempfile creation. I just assume that there is already a newly created tempfile and we use it like any other file (or stream).
Is it guaranteed that all "written" data can be "read" again? Or might some data be hanging around in the buffer of the (not yet flushed) write process? Maybe Boolector is already flushing at the end of its method, then we do not need to flush by ourselves.

Why am I asking this? Because the ending of the String from dumping a single variable (see the added JUnit test) contains some strange chars where I cannot detect a reason for. I have seen at least the String endings `"\0"`, `"der;"`, and `""` for the same (!) variable, even while dumping the same node several times within one program execution. A possible reason for this might be that some content of the file, such as its last line, is not yet written to file before the file is read into a String.

Reply:
You were right, a flush of the file before use fixes the problem.

Why doesn't C guarantee that a newly used file is empty though?!

Reply:
Now I am confused. I thought about flushing **after** calling `boolector_dump_smt2_node`, not **before**. I assumed that the filepointer points to the start of the opened file and Boolector starts to write there.
Now the control flow is such that:
- we create a file
- we flush the file (why? There is no content.)
- Boolector writes to the file
- we do not flush the file (I really expected that flush is needed here!)
- JNI reads the file into a String.
Strange. I will test and report whether it works.

Reply:
My bad, i misunderstood the behavior of `fflush()`, you are right of course.

Reply:
I just recompiled commits [Fix SigSegv](https://github.com/sosy-lab/java-smt/pull/185/commits/68cb0d564394ab020f40e8c612fac8b1a5ba9530) and [Fixed compile warnings](https://github.com/sosy-lab/java-smt/pull/185/commits/ca096fe7b07396eb3707d1c6956751714670b510) and tested them both.
The problem seems to be present in the first, but not in the later, thats why my improper flush "worked", it was already fixed, at least on my system.
Can you confirm?